keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,883 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposed enhancement adds a new `VolumeSource` to Kubernetes that supports OCI images and/or OCI artifacts.
+This allows users to package files and share them among containers in a pod without including a shell in the image,
+thereby reducing vulnerabilities and simplifying image creation.
+
+While OCI images are well-supported by Kubernetes and CRI,
+extending support to OCI artifacts involves recognizing additional media types, implementing custom lifecycle management,
+and ensuring appropriate validation and security measures.
+
+## Motivation
+
+Currently, mounting files from OCI images requires the inclusion of a shell, even when the sole purpose of the image is to deliver files
+to be shared among containers in a pod. This not only increases the image size unnecessarily but also introduces potential security vulnerabilities.
+By supporting OCI images and artifacts directly as a `VolumeSource`, we can streamline the process and enhance security.
+
+### Goals
+
+- Introduce a new `VolumeSource` type that allows mounting OCI images and/or artifacts.
+- Simplify the process of sharing files among containers in a pod.
+- Reduce the need for including a shell in images, thereby minimizing security risks.
+- Reuse existing logic from ConfigMaps for managing file locations, garbage collection, and updates.
+
+### Non-Goals
+
+- This proposal does not aim to replace existing `VolumeSource` types.
+- This proposal does not address other use cases for OCI objects beyond file sharing among containers in a pod.
+
+## Proposal
+
+We propose to add a new `VolumeSource` that supports OCI images and/or artifacts. This `VolumeSource` will allow users to mount an OCI object
```

Comment:
Can you use Volume Populators?
One use case for Volume Populator is VM image from KubeVirt which is similar to your use case:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1495-volume-populators#vm-images

Reply:
also leans into the RHCOS pattern

Reply:
Hmm, the Volume Populators are definitely related, but seem a bit more block-device oriented to me?  Whereas this is much more filesystem oriented.

Reply:
I agree that's a more filesystem oriented approach, not exactly sure how volume populators would fit.

Reply:
Secrets and ConfigMaps are technically hardcoded volume populators for empty dir (implementation wise).  An ephemeral CSI driver can have a volume populator, right?